vllm/core/block_manager_v2.py

```diff
@@ -227,17 +228,89 @@ def fork(self, parent_seq: Sequence, child_seq: Sequence) -> None:
 
     def can_swap_in(self, seq_group: SequenceGroup,
                     num_lookahead_slots: int) -> bool:
-        return False
+        """
+        We go through all sequence in seq group to get their number of blocks 
+        touched and sum them up to see whether there is enough memory to swap in
+        """
+        num_touched_blocks = 0
+        for seq in seq_group.get_seqs(status=SequenceStatus.SWAPPED):
+            block_table = self.block_tables[seq.seq_id]
+            num_touched_blocks += (
+                block_table.get_num_blocks_touched_by_append_slots(
+                    token_ids=seq.get_token_ids(),
+                    num_lookahead_slots=num_lookahead_slots,
+                ))
+        num_free_blocks = self.block_allocator.get_num_free_blocks(Device.GPU)
+        return num_free_blocks - num_touched_blocks >= self.watermark_blocks
 
     def swap_in(self, seq_group: SequenceGroup,
                 num_lookahead_slots: int) -> Dict[int, int]:
-        raise NotImplementedError
+        mapping: Dict[Block, Block] = {}
+        for seq in seq_group.get_seqs(status=SequenceStatus.SWAPPED):
+            new_block_table = BlockTable(
+                block_size=self.block_size,
+                block_allocator=self.block_allocator,
+            )
+            block_table = self.block_tables[seq.seq_id]
+
+            for cpu_block in block_table:
+                if cpu_block in mapping:
+                    gpu_block = mapping[cpu_block]
+                    self.block_allocator.increase_ref_count(
+                        Device.GPU, gpu_block.block_id())
+                else:
+                    gpu_block = new_block_table.append_by_blocks(
+                        token_ids=cpu_block.token_ids(), device=Device.GPU)
+                    mapping[cpu_block] = gpu_block
+                self.block_allocator.free(cpu_block)
+            self.block_tables[seq.seq_id] = new_block_table
+
+        # NOTE: since the memory operation in physical blocks need the
```

Comment:
The device aware block manager could swap to disk in addition to CPU in the future. So we don't want this layer to know details on how to swap blocks -- it should invoke the device aware block allocator for this, and simply get a new mapping.

In a related way, this layer should not know about refcounts -- it should be pushed to lower layers.

Reply:
Thanks for your comment, pushing the logic down to the `DeviceAwareBlockAllocator` class, PTAL.

Reply:
Can we push even more into the lower layer?

```python3
class BlockManagerV2:
    def swap_in(self, seq_group, lookahead):
        for seq in seqs:
            self.block_tables[seq.seq_id].swap(source_device=Device.CPU, destination_device=Device.GPU)
            # NOTE: scheduler must now retrieve copy-on-writes separately
            self.append_slots(seq, lookahead)

        # Once the BlockManagerV1 implementation is deleted, we can move this
        # `get_and_reset_swaps` call outside of `swap_in`/`swap_out`. Then the
        # scheduler can make a single call to get all swaps / all copy-on-writes for
        # the batch.
        all_swaps = self._allocator.get_and_reset_swaps()
        return all_swaps

    def swap_out(self, seq_group):
        for seq in seqs:
            self.block_tables[seq.seq_id].swap(source_device=Device.GPU, destination_device=Device.CPU)
        # See https://github.com/vllm-project/vllm/blob/main/vllm/core/block/common.py#L145-L160
        # for design inspiration
        all_swaps = self._allocator.get_and_reset_swaps()
        return all_swaps
```